projects/etos_suite_runner/src/etos_suite_runner/lib/suite.py

```diff
@@ -318,6 +319,67 @@ def sub_suite_environments(self) -> Iterator[dict]:
             self._record_exception(exc)
             raise exc
 
+    def _sub_suite_environments_from_kubernetes(self) -> Iterator[dict]:
+        """All sub suite environments from Kubernetes.
+
+        Each sub suite environment is an environment for the sub suites to execute in.
+        """
+        environments = []
+        timeout = time.time() + self.etos.config.get("WAIT_FOR_ENVIRONMENT_TIMEOUT")
+        while time.time() < timeout:
+            time.sleep(5)
+            for environment in self.params.environments:
+                if environment.spec.sub_suite_id in environments:
+                    continue
+
+                # Send eiffel event for the ETR.
+                event = EiffelEnvironmentDefinedEvent()
+                event.meta.event_id = environment.metadata.name
+                url = f"{os.getenv('ETOS_API')}/v1alpha/testrun/{environment.metadata.name}"
+                self.etos.events.send(
+                    event,
+                    {"CONTEXT": self.test_suite_started_id},
+                    {"name": environment.spec.name, "uri": url},
+                )
+
+                environments.append(environment.spec.sub_suite_id)
+                executor = environment.spec.executor
+                yield {
+                    "executor": executor,
+                    "id": environment.spec.sub_suite_id,
+                    "name": environment.spec.name,
+                }
+            status = self.params.get_status()
+            if status.get("status") == "FAILURE":
```

Comment:
If neither of these two if-conditions is True and the while-timeout hasn't expired yet, the loop will continue.
It will yield same environments with the same events. Not a problem? If not, please comment why.

Reply:
It will not yield the same environments since all environments are added to a list on line 345 and checked on line 332.
The point of this loop is to continue until either of these two if-conditions are True.

Not sure what you want me to comment in the code for that.